src/platform/windows/mod.rs

```diff
@@ -0,0 +1,1482 @@
+                            selection_results.push(OsIpcSelectionResult::DataReceived(completion_key as i64, msg.0, msg.1, msg.2));
+                        } else {
+                            dd!("[# {:?}] receiver {:?} -- no message!", *self.iocp, *receiver.handle);
+                            break;
+                        }
+                    }
+
+                    // if it wasn't closed, kick off a read
+                    try!(reader.start_read());
+
+                    // we may have already been closed, or the read resulted in us being closed
+                    if reader.closed {
+                        dd!("[# {:?}] receiver {:?} -- now closed!", *self.iocp, *receiver.handle);
+                        selection_results.push(OsIpcSelectionResult::ChannelClosed(completion_key as i64));
+                        remove_index = index;
+                    }
+
+                    break;
+                }
+
+                // if we were told to remove this receiver, do so
+                if remove_index < receivers.len() {
+                    receivers.swap_remove(remove_index);
+                }
+
+                // if we didn't dequeue at least one complete message -- we need to loop through GetQueuedCS again;
+                // otherwise we're done.
+                if !selection_results.is_empty() {
+                    break;
+                }
+            }
+
+            dd!("select() -> {} results", selection_results.len());
+            Ok(selection_results)
+        }
+    }
+}
+
+impl OsIpcSelectionResult {
+    pub fn unwrap(self) -> (i64, Vec<u8>, Vec<OsOpaqueIpcChannel>, Vec<OsIpcSharedMemory>) {
+        match self {
+            OsIpcSelectionResult::DataReceived(id, data, channels, shared_memory_regions) => {
+                (id, data, channels, shared_memory_regions)
+            }
+            OsIpcSelectionResult::ChannelClosed(id) => {
+                panic!("OsIpcSelectionResult::unwrap(): receiver ID {} was closed!", id)
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct OsIpcSharedMemory {
+    handle: WinHandle,
+    ptr: *mut u8,
+    length: usize,
+}
+
+unsafe impl Send for OsIpcSharedMemory {}
+unsafe impl Sync for OsIpcSharedMemory {}
+
+impl Drop for OsIpcSharedMemory {
+    fn drop(&mut self) {
+        unsafe {
+            if self.handle.is_valid() {
+                kernel32::UnmapViewOfFile(self.ptr as LPVOID);
+            }
+        }
+    }
+}
+
+impl Clone for OsIpcSharedMemory {
+    fn clone(&self) -> OsIpcSharedMemory {
+        let handle = dup_handle(*self.handle).unwrap();
+        OsIpcSharedMemory::from_handle(handle, self.length).unwrap()
+    }
+}
+
+impl PartialEq for OsIpcSharedMemory {
+    fn eq(&self, other: &OsIpcSharedMemory) -> bool {
+        unsafe {
+            self.length == other.length &&
+            memcmp(self.ptr as *const _, other.ptr as *const _, self.length) == 0
```

Comment:
Well, that test is not actually about cloned mappings: what it checks for (among other things) is that a SHM region sent to ourselves (that is, to a receiver in the same process as the sender) will reuse the mapping, rather than establishing a new one -- and you are basically cheating your way around the fact that the test actually exposes different behaviour of your implementation, by also implementing a misbehaving equality check to hide that difference... Maybe we should add a separate test that checks whether the equality function actually behaves the way it's supposed to :-P

Having said that, I don't know whether the self-sending behaviour is actually considered an important property, that really needs to be tested for...